Scripts/RemediationScripts/Remediate-ConfigureConditionalAccessPolicyForPIM.ps1

```diff
@@ -279,48 +296,73 @@ function Configure-ConditionalAccessPolicyForPIM
 
     # No file path provided as input to the script. Fetch currently configured Conditional Access (CA) policy in the subscription.
     if ([String]::IsNullOrWhiteSpace($FilePath))
     {
         $policyAssignments = Get-AzRoleManagementPolicyAssignment -Scope "subscriptions/$($SubscriptionId)"
-
+        $policyDetailsCollection = Get-AzRoleManagementPolicy -Scope "subscriptions/$($SubscriptionId)"
+        $roleDefinitions = $policyAssignments | Select-Object -ExpandProperty RoleDefinitionDisplayName
+        
         if (-not $RoleName)
-        {
-            $adminRoles = @("Owner", "Contributor", "User Access Administrator")
+        {   
+            # Remove excluded roles from the role definition collection.
+            $criticalRoles = $roleDefinitions | Where-Object { $_ -notin $excludedRoles }
         }
         else
         {
-            $adminRoles = @($RoleName)
+            # Validate if RoleName is a valid for the particular subscription (role name can be subscription specific for custom roles).
+            if (-not $roleDefinitions.Contains($RoleName))
+            {
+                Write-Host "The role ($($RoleName)) provided does not exist for the subsciption ($($SubscriptionId)). Exiting..."
+                return
+            }
+
+            $criticalRoles = @($RoleName)
         }
 
+        # Fetch all the CA policies and check if for all the critical roles, policy has been assigned.
         $nonCompliantRoles = @()
+        $nonCompliantRolesCount = 0
         $configuredPolicyDetails = New-Object -TypeName PSObject
-        $adminRoles | ForEach-Object {
+
+        # The Hash Table below will contain key-value pair of the form (role Id, mfa Rule) if MFA is enabled for that role Id. 
+        $mfaEnabledRoles = @{}
+        $criticalRoles | ForEach-Object {
             $role = $_
             $policyId = $policyAssignments | Where-Object {$_.RoleDefinitionDisplayName -contains $role}
             $roleId = $policyId.PolicyId.Split('/')
         
-            $policyDetails = Get-AzRoleManagementPolicy -Scope "subscriptions/$($SubscriptionId)" -Name $roleId[-1]
+            $policyDetails = $policyDetailsCollection | Where-Object { $_.Name -eq $roleId[-1] }
             $configuredPolicyDetail = $policyDetails.EffectiveRule | Where-Object {($_.claimValue -eq "c1" -and $_.IsEnabled -eq $true) -or ($_.claimValue -eq "urn:microsoft:req1" -and $_.IsEnabled -eq $true)}
-
             if ([String]::IsNullOrWhiteSpace($configuredPolicyDetail))
             {
                 $nonCompliantRoles += $role
+                $nonCompliantRolesCount += 1
                 $nonCompliantRolesStr = $nonCompliantRoles -join ','
             }
+
+            $targetRuleForMFA = $policyDetails.EffectiveRule | Where-Object { $_.Id -eq 'Enablement_EndUser_Assignment' -and $_.RuleType -eq 'RoleManagementPolicyEnablementRule' }
```

Comment:
NIT: Please add comment, MFA check needed, so that in future we will know the context.